misc/amia_paper_2024/stats.py

```diff
@@ -0,0 +1,160 @@
+"""AMIA paper 2024: Statistics
+
+todo's (minor)
+ - get_dataset_with_mods(): Messy to have fetching of dataset and also modification in 1 function. This adds client IP.
+
+Resources
+- GH issue: https://github.com/jhu-bids/TermHub/issues/631
+- Paper: https://livejohnshopkins-my.sharepoint.com/:w:/g/personal/sgold15_jh_edu/EXbuxwBsb0pPgFfgPYjR6qgBlJ1gmpqsMPYJ3vn6Y_HnYw?e=4%3ApEDbMe&fromShare=true&at=9&CID=d4663c40-1ebe-db59-8cf5-2bc091d45692"""
+import os
+import sys
+from datetime import datetime
+from pathlib import Path
+from typing import Callable, Dict, List, Union
+
+import pandas as pd
+from sqlalchemy import RowMapping
+
+from backend.db.utils import get_db_connection, sql_query
+
+THIS_DIR = Path(os.path.dirname(__file__))
+PROJECT_ROOT = THIS_DIR.parent.parent
+sys.path.insert(0, str(PROJECT_ROOT))
+from backend.routes.db import usage_query
+
+
+INDIR = THIS_DIR / 'input'
+OUTDIR = THIS_DIR / 'output'
+USAGE_JOINED_CSV_PATH = INDIR / 'api_runs_w_groups_filtered.csv'
+USAGE_UNJOINED_CSV_PATH = INDIR / 'api_runs.csv'
+OUT_CSV_PATH_ALL = OUTDIR / 'summary.csv'
+OUT_CSV_DEV_UNFILTERED_PATH = OUTDIR / 'summary_dev_unfiltered.csv'
+OUT_CSV_DEV_FILTERED_PATH = OUTDIR / 'summary_dev_filtered.csv'
+
+
+def setup():
+    """Setup any environmental prerequisites"""
+    if not OUTDIR.exists():
+        OUTDIR.mkdir()
+
+
+def api_runs_query(verbose=False):
+    """Get all records from api_runs table"""
+    t0 = datetime.now()
+    with get_db_connection() as con:
+        data: List[RowMapping] = sql_query(con, """SELECT DISTINCT * FROM public.api_runs r""")
+    data: List[Dict] = [dict(d) for d in data]
+    if verbose:
+        print(f'api_runs_query(): Fetched {len(data)} records in n seconds: {(datetime.now() - t0).seconds}')
+    return data
+
+
+def get_dataset_with_mods(func: Callable, path: Union[str, Path], use_cache=False, verbose=True) -> pd.DataFrame:
+    """Get data to be analyzed: JOIN of api_runs and apiruns_grouped
+
+    Mods: Adds client IP column. Filters out null call groups."""
+    if not use_cache:
+        data: List[Dict] = func(verbose)
+        df = pd.DataFrame(data)
+    else:
+        # Fetch data
+        if not os.path.exists(path):
+            # Fetch
+            data: List[Dict] = func(verbose)
+            df = pd.DataFrame(data)
+            # Save
+            t0 = datetime.now()
+            df.to_csv(path, index=False)
+            if verbose:
+                print(f'Wrote {len(df)} records to {path} in n seconds: {(datetime.now() - t0).seconds}')
+        else:
+            # Read
+            t0 = datetime.now()
+            df = pd.read_csv(path)
+            if verbose:
+                print(f'Read {len(df)} records from {path} in n seconds: {(datetime.now() - t0).seconds}')
+    # Mods
+    df['client_ip'] = df['client'].apply(lambda x: x.split(':')[0])
```

Comment:
## IP address stuff
The tot IP addresses without dev IPs is 234, and with dev IPs is 230. @Sigfried You may wonder why the difference is >2. It's because of localhost:

The 4 'dev ips':
- '174.99.54.40'
- 'no forward; 127.0.0.1 (no loc, err 106)'
- '127.0.0.1'
- '216.164.48.98'

<details><summary>Additional details</summary>
<p>

Code I used to get this diff:
```py
  # developer_ips = {'Siggie': '216.164.48.98', 'Joe': '174.99.54.40'}
  ip = set(df_apiruns['client_ip'].unique())
  ip_nodev = set(preprocess_dev_ips(df_apiruns, verbose)['client_ip'].unique())
  diff = ip.difference(ip_nodev)
```

Note that `client_ip` is a new column. This is the result of the operation:
`df['client_ip'] = df['client'].apply(lambda x: x.split(':')[0])`

If I did the diff on `client`, it would be:
- 'no forward; 127.0.0.1 (no loc, err 106)'
- '216.164.48.98:55408; 169.254.130.3'
- '127.0.0.1'
- '174.99.54.40: Raleigh, no region'
- '216.164.48.98:61752; 169.254.130.6'
- '216.164.48.98: Bethesda, Maryland'
- '174.99.54.40: Raleigh, North Carolina'
- '216.164.48.98:57532; 169.254.130.6'
- '216.164.48.98: Takoma Park, Maryland'
- '216.164.48.98: Lake Ridge, Virginia'
- '216.164.48.98:55659; 169.254.130.6'
- '216.164.48.98: Wheaton, Maryland'
- '216.164.48.98: Takoma Park, no region'

</p>
</details>